README.md
oauth-login-implementation-examples-using-rauth-via-bottle.py
=============================================================

OAuth Login Implementation Examples using rauth via Bottle.py

Prerequisites:
==============
**System Packages:**
* For Python packages:
  * `python-setuptools`
  * `pip`
  * `virtualenv`

**Python Packages:**
* `bottle`
* `rauth`

Installing the prerequisites:
=============================
1. Install `python-setuptools`: `sudo apt-get install python-setuptools`
2. Install `pip`: `sudo easy_install -U pip`
3. Install `virtualenv`: `sudo pip install -U virtualenv`
4. Clone the repository: `git clone https://github.com/ejelome/oauth-login-implementation-examples-using-rauth-via-bottle.py`
5. Go to the directory: `cd oauth-login-implementation-examples-using-rauth-via-bottle.py`
6. Create a virtualenv: `virtualenv venv`
7. Activate virtualenv: `source venv/bin/activate`
8. Install remaining prerequisites (`bottle` and `rauth`): `pip install -r requirements.txt`

Modify the `hosts` file:
========================
In order to make the OAuth APIs work in our local machine, we need to modify the hosts file and pretend we're visiting an online URI:
* When testing Facebook and Twitter examples, make sure that the local address (`127.0.0.1`) points to `mydomain.tld`
* When testing the Google example, the `mydomain.tld` won't work since Google seem to know that it's an invalid URI
* We can solve these issues by setting 2 aliases for the `127.0.0.1` address

1. Open the hosts file: `sudo gedit /etc/hosts`
2. Add the following line: `127.0.0.1 mydomain.tld`

Running the program:
====================
1. Execute one of the following:
```
python oauth2-facebook.py # Facebook
python oauth1-twitter.py # Twitter
python oauth2-google.py # Google
```

2. Open a browser and go to: [http://localhost:8000](http://localhost:8000)
3. To Login:
   * Click *Log in using xyz*, where xyz is either Facebook, Twitter, or Google
   * Or go to: [http://localhost:8000/login](http://localhost:8000/login)

Notes:
======
* More detailed information are included in each of the executable files (there are also in the `config.py` file)
* You can change the OAuth API settings on `config.py` file
* You can change other configurations on the `config.py` file
* Coding conventions are based on PEP 8 to make sure that you're also seeing good code examples
* The examples were done as minimal as possible so you can get a better understanding of OAuth and how simple it is
* Happy coding! :)

Recent Changes:
===============
See: [CHANGELOG](https://github.com/ejelome/oauth-login-implementation-examples-using-rauth-via-bottle.py/blob/master/CHANGELOG)




CHANGELOG
2010-01-10 EJelome <ejelome@gmail.com>
* README.md: Transferred Recent Changes' logs to this file
  - 2014-01-09:
    : Added MIT LICENSE and .gitignore files
    : Added python-setuptools on the prerequisites and installing prerequisites
  - 2014-01-10:
    : Re-tested and updated installation and usage instructions


config.py
#!/usr/bin/python2.7
# -*- coding: utf-8 -*-

"""config.py: Default configuration."""

# Server:
SERVER = 'wsgiref'
HOST = 'localhost'
PORT = 8000

# Meta:
# Note on making it work in localhost:
# * Open a terminal, then do:
#   - sudo gedit /etc/hosts
# * Enter the desired localhost alias for 127.0.0.1:
#   - (e.g. 127.0.0.1 mydomain.tld)
# * Don't forget to save the file :)
BASE_URI = 'http://mydomain.tld'
GOOGLE_BASE_URI = 'http://localhost' # Google doesn't seem to accept
                                     # non-working urls, but accepts localhost

# Facebook:
FACEBOOK_CLIENT_ID = '545723862164937'
FACEBOOK_CLIENT_SECRET = '5407608c6e9d10b297ba77a4de18f75f'

# Twitter:
TWITTER_CLIENT_ID = '07gvoVMGyyul4m5bKkJuOA'
TWITTER_CLIENT_SECRET = 'PUOGZog1IhxsOQVojrwVgjHgSlHaVRibcpVsP8Axk0'

# Google:
GOOGLE_CLIENT_ID = '954851881468.apps.googleusercontent.com'
GOOGLE_CLIENT_SECRET = 'HTvpTPPUGQfWByCaonsC4EMa'


oauth1-twitter.py
#!/usr/bin/python2.7
# -*- coding: utf-8 -*-

"""oauth1-twitter.py: Implementing OAuth1 Login for Twitter."""

# Guidelines:
# Note: Twitter uses OAuth1, not OAuth2
# 1. Create New App at: https://dev.twitter.com/apps/new
# 2. On the Application Details, specify the following:
#    * Name: (e.g. My XYZ App)
#    * Description: (e.g. Test App for My App)
#    * Website: (e.g. mydomain.tld)
#    * Callback URL: (e.g. http://mydomain.tld/success)
#    * Click the "Yes, I agree" checkboxk
#    * On the CAPTCHA, enter the image texts
#    * Click the Create your Twitter application
# 3. On the https://dev.twitter.com/apps/<app_id>/show, do the following:
#    * Get the Consumer key: (e.g. 07gvoVMGyyul4m5bKkJuOA)
#      - This is your consumer_key
#    * Get the Consumer secret:
#      - (e.g. PUOGZog1IhxsOQVojrwVgjHgSlHaVRibcpVsP8Axk0)
#      - This is your consumer_secret
#    * Under "Your access token", click Create my access token

import bottle as app
import rauth

import config

oauth1 = rauth.OAuth1Service
twitter = oauth1(
    consumer_key=config.TWITTER_CLIENT_ID,
    consumer_secret=config.TWITTER_CLIENT_SECRET,
    name='twitter',
    request_token_url='https://api.twitter.com/oauth/request_token',
    access_token_url='https://api.twitter.com/oauth/access_token',
    authorize_url='https://api.twitter.com/oauth/authorize',
    base_url='https://api.twitter.com/1.1/'
)
request_token, request_token_secret = twitter.get_request_token()
authorize_url = twitter.get_authorize_url(request_token)

@app.route('/')
def index():
    return '<a href="/login">Log in using Twitter</a>'

@app.route('/login<:re:/?>')
def login():
    app.redirect(authorize_url)

@app.route('/success<:re:/?>')
def login_success():
    oauth_token = app.request.params.get('oauth_token')
    oauth_verifier = app.request.params.get('oauth_verifier')
    session = twitter.get_auth_session(
        request_token,
        request_token_secret,
        method='POST',
        data=dict(oauth_verifier=oauth_verifier)
    )
    params = {}
    # For complete list of parameters for verify_credentials,
    # go to: https://dev.twitter.com/docs/api/1.1/get/account/verify_credentials
    # Note: This same technique applies to other json_path, just got to its page
    #       and the parameters can be found there

    json_path = 'account/verify_credentials.json'
    session_json = session.get(json_path, params=params, verify=True).json()
    # For complete list of json_path,
    # go to: https://dev.twitter.com/docs/api/1.1
    return 'Welcome {name}!'.format(**session_json)

app.run(
    server=config.SERVER,
    port=config.PORT,
    host=config.HOST
)


oauth2-facebook.py
#!/usr/bin/python2.7
# -*- coding: utf-8 -*-

"""oauth2-facebook.py: Implementing OAuth2 Login for Facebook."""

# Guidelines:
# 1. Register your account as a developer at https://developers.facebook.com/.
# 2. Create New App at: https://developers.facebook.com/apps/.
# 3. On https://developers.facebook.com/apps/<app_id>/summary, do the following:
#    * Get App ID: (e.g. 545723862164937)
#      - This is your client_id
#    * Get App Secret: (e.g. 5407608c6e9d10b297ba77a4de18f75f)
#      - This is your client_secret
#    * On App Domains: (e.g. mydomain.tld)
#    * On Sandbox Mode:
#      - Click Disabled to let non-developer accounts test the app
#    * Under the "Select how your app integrates with Facebook",
#      click the "Website with Facebook Login"
#    * Click Save Changes

import bottle as app
import rauth

import config

oauth2 = rauth.OAuth2Service
facebook = oauth2(
    client_id=config.FACEBOOK_CLIENT_ID,
    client_secret=config.FACEBOOK_CLIENT_SECRET,
    name='facebook',
    authorize_url='https://graph.facebook.com/oauth/authorize',
    access_token_url='https://graph.facebook.com/oauth/access_token',
    base_url='https://graph.facebook.com/'
)
redirect_uri = '{uri}:{port}/success'.format(
    uri=config.BASE_URI,
    port=config.PORT
)

@app.route('/')
def index():
    return '<a href="/login">Log in using Facebook</a>'

@app.route('/login<:re:/?>')
def login():
    params = dict(
        scope='read_stream',
        response_type='code',
        redirect_uri=redirect_uri
    )
    url = facebook.get_authorize_url(**params)

    app.redirect(url)

@app.route('/success<:re:/?>')
def login_success():
    code = app.request.params.get('code')
    session = facebook.get_auth_session(
        data=dict(
            code=code,
            redirect_uri=redirect_uri
        )
    )
    session_json = session.get('me').json()
    # For complete list of keys,
    # go to: https://developers.facebook.com/docs/reference/api/user/
    return 'Welcome {name}!'.format(**session_json)

app.run(
    server=config.SERVER,
    port=config.PORT,
    host=config.HOST
)


oauth2-google.py
#!/usr/bin/python2.7
# -*- coding: utf-8 -*-

"""oauth2-google.py: Implementing OAuth2 Login for Google."""

# Guidelines for rauth via Google API:
# 1. Create project at:
#    https://cloud.google.com/console?redirected=true#/project?redirected=true
# 2. On the New Project, specify the following:
#    * Project name: (e.g. My XYZ App)
#    * Project ID: (e.g. my-xyz-app)
# 3. Go to: https://code.google.com/apis/console/#:access
#    * Click Edit branding information... and specify the following:
#      * Product name: (e.g. My XYZ App)
#      * Click Update
#    * Click Create another client ID... and specify the following:
#      * On Application type, select Web application (default)
#      * Under Your site or hostname:
#        * Click more options:
#          * Under the Authorized Redirect URIs, type the following:
#            - http://localhost:8000/success
#          * Under the Authorized JavaScript Origins, type the following:
#            - http://localhost:8000
#          * Click Create client ID
#   * Under the Client ID for web applications, get the following:
#     * Client ID: (e.g. 954851881468.apps.googleusercontent.com)
#     * Client secret: (e.g. HTvpTPPUGQfWByCaonsC4EMa)

import json

import bottle as app
import rauth

import config

oauth2 = rauth.OAuth2Service
google = oauth2(
    client_id=config.GOOGLE_CLIENT_ID,
    client_secret=config.GOOGLE_CLIENT_SECRET,
    name='google',
    authorize_url='https://accounts.google.com/o/oauth2/auth',
    access_token_url='https://accounts.google.com/o/oauth2/token',
    base_url='https://accounts.google.com/o/oauth2/auth',
)
redirect_uri = '{uri}:{port}/success'.format(
    uri=config.GOOGLE_BASE_URI,
    port=config.PORT
)

@app.route('/')
def index():
    return '<a href="/login">Log in using Google</a>'

@app.route('/login<:re:/?>')
def login():
    params = dict(
        scope='email profile',
        response_type='code',
        redirect_uri=redirect_uri
    )
    url = google.get_authorize_url(**params)

    app.redirect(url)

@app.route('/success<:re:/?>')
def login_success():
    code = app.request.params.get('code')
    session = google.get_auth_session(
        data=dict(
            code=code,
            redirect_uri=redirect_uri,
            grant_type='authorization_code'
        ),
        decoder=json.loads
    )
    json_path = 'https://www.googleapis.com/oauth2/v1/userinfo'
    session_json = session.get(json_path).json()
    return 'Welcome {name}!'.format(**session_json)
    # Unfortunately, there seem to be no reference for complete list of keys,
    # but here are the complete json keys returned by the scope: email profile:
    # * email
    # * family_name
    # * gender
    # * given_name
    # * id
    # * link
    # * locale
    # * name
    # * picture

app.run(
    server=config.SERVER,
    port=config.PORT,
    host=config.HOST
)
